This repository houses code developed during my tenure at UPenn as a bioinformatician in the [Faryabi Lab](https://faryabilab.com/). The code found here deals with the ingestion, analysis, and visualization of Next-Generation Sequencing (NGS) data, as well as utility functions to do a wide array of tasks common to bioinformatics and data analysis in Python and R. There are also a wide array of scripts for bioinformaticians working in a UNIX environment. 
********
### **Structure**: 
```
CodeWorkbench/
├── 3C-methods/     # Hi-C, HiChIP, Micro-C analysis
├── compbio/        # DE testing, super enhancers, 3D hubs
├── imaging/        # Code for image analysis (currently FISH and ORCA)
├── scripts/        # General-purpose plotting and utility scripts
  ├──  convert/     # File format converters for common bioinformatics filetypes
  ├── misc/         # Miscellaneous scripts and helpers
  ├── DepMap/       # Tools for Cancer Dependency Map data
```

#### Other repositories
Not all of my code/projects at Penn are housed here. They can be found at:
* https://github.com/faryabiLab/dockerize-workflows - Containerized, version controlled, and reproducible Next Generation Sequencing (NGS - RNAseq, ChIPseq, ATACseq) processing pipelines.
    * https://github.com/Nuwah12/Demo-RNAseq-Pipeline - Simple step-by-step RNA-seq processing pipeline written in Bash made as a teaching tool for non-computational colleagues.
* https://github.com/Nuwah12/GENOVA2 - A **fork** of the popular 3C data visualization R package [GENOVA](https://github.com/robinweide/GENOVA), with added options for applying normalization to contact matrices.
* https://github.com/Nuwah12/EBV_Project - Code related to a project surrounding the interaction of an MCL cell line's genome with the Epstein-Barr Virus episome (coming soon, currently under construction).
 
In addition, my [personal website](https://nuwah12.github.io/) includes more information on my work and career. 

****
