This repository houses code developed during my tenure at UPenn in the [Faryabi Lab](https://faryabilab.com/). The code found here deals with the ingestion, analysis, and visualization of Next-Generation Sequencing (NGS) data, as well as utility functions to do a wide array of tasks common to bioinformatics and data analysis in Python and R, as well as shell scripts for working in a UNIX environment. 
********
### **Structure**: 
#### 
`3C-methods/` - analysis scripts and other code for working with 3C-based assays (Hi-C, HiChIP, MicroC, etc.) \
`compbio/` - scripts for differential expression testing, super enhancer identification, and 3D hub formation. \
`ORCA/` - scripts for analysis of ORCA data and simulated ORCA traces. *All code related to polymer simulations is here*.\
`scripts/` - miscellaneous scripts for file conversion, data preprocessing, plotting/data visualization, etc.

#### Other repositories
Not all of my code/projects at Penn are housed here. They can be found at:
* https://github.com/faryabiLab/dockerize-workflows - Containerized and reproducible NGS processing pipelines.
    * https://github.com/Nuwah12/Demo-RNAseq-Pipeline - Simple step-by-step RNA-seq processing pipeline made as a teaching tool for non-computational coleagues.
* https://github.com/Nuwah12/GENOVA2 - A **fork** of the popular 3C data visualization R package [GENOVA](https://github.com/robinweide/GENOVA), with added options for applying normalization to contact matrices.
* https://github.com/Nuwah12/EBV_Project - Code related to a project surrounding the interaction of an MCL cell line's genome with the Epstein-Barr Virus episome (coming soon, currently under construction).
 
